src/Compilers/CSharp/Portable/Errors/CustomObsoleteDiagnosticInfo.cs

```diff
@@ -0,0 +1,117 @@
+        {
+            // dev11 had a bug in this area (i.e. always produce a warning when there's no message) and we have to match it.
+            if (data.Message is null)
+            {
+                return inCollectionInitializer ? ErrorCode.WRN_DeprecatedCollectionInitAdd : ErrorCode.WRN_DeprecatedSymbol;
+            }
+
+            return (data.IsError, inCollectionInitializer) switch
+            {
+                (true, true) => ErrorCode.ERR_DeprecatedCollectionInitAddStr,
+                (true, false) => ErrorCode.ERR_DeprecatedSymbolStr,
+                (false, true) => ErrorCode.WRN_DeprecatedCollectionInitAddStr,
+                (false, false) => ErrorCode.WRN_DeprecatedSymbolStr
+            };
+        }
+
+        private static object[] GetArguments(Symbol obsoletedSymbol, ObsoleteAttributeData obsoleteAttributeData)
+        {
+            var message = obsoleteAttributeData.Message;
+            if (message is object)
+            {
+                return new object[] { obsoletedSymbol, message };
+            }
+            else
+            {
+                return new object[] { obsoletedSymbol };
+            }
+        }
+
+        public override string MessageIdentifier
+        {
+            get
+            {
+                var id = Data.DiagnosticId;
+                if (!string.IsNullOrEmpty(id))
+                {
+                    return id;
+                }
+
+                return base.MessageIdentifier;
+            }
+        }
+
+        public override DiagnosticDescriptor Descriptor
+        {
+            get
+            {
+                var baseDescriptor = base.Descriptor;
+                var diagnosticId = Data.DiagnosticId;
+                var urlFormat = Data.UrlFormat;
+                if (diagnosticId is null && urlFormat is null)
+                {
+                    return baseDescriptor;
+                }
+
+                var id = MessageIdentifier;
+                var helpLinkUri = baseDescriptor.HelpLinkUri;
+
+                if (urlFormat is object)
+                {
+                    try
+                    {
+                        helpLinkUri = string.Format(urlFormat, id);
+                    }
+                    catch
+                    {
+                        // TODO: should we report a meta-diagnostic of some kind when the string.Format fails?
+                        // also, should we do some validation of the 'UrlFormat' values provided in source to prevent people from shipping
+                        // obsoleted symbols with malformed 'UrlFormat' values?
+                    }
+                }
+
+                var customTags = ArrayBuilder<string>.GetInstance();
+                foreach (var tag in baseDescriptor.CustomTags)
+                {
+                    customTags.Add(tag);
+                }
+                customTags.Add(WellKnownDiagnosticTags.CustomObsolete);
```

Comment:
>WellKnownDiagnosticTags.CustomObsolete [](start = 31, length = 38)

Should we add this when ```diagnosticId is null```?

Reply:
Good question. I don't feel certain whether this tag is actually needed. I added it to hint to `Test.Utilities.DiagnosticDescription` that this diagnostic ID is user-defined, but I'm currently checking `is CustomObsoleteDiagnosticInfo { Data: { DiagnosticId: object _ } }` instead in there.

If I made the change you're implying (add the custom tag iff a custom DiagnosticId is present), we would be able to rely on that, which seems simpler.